Conveyor belt component.

michalpelka opened the pull request:
Addresses: #364

jhanca-robotecai commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+
+        AZ::Matrix3x3 rotationMatrix = AZ::Matrix3x3::CreateFromColumns(v1, v2, v3);
+        AZ_Assert(rotationMatrix.IsOrthogonal(0.001f), "Rotation matrix is not orthogonal");
+        rotationMatrix.Orthogonalize();
```
Is this needed?

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+#include <AzFramework/Physics/RigidBodyBus.h>
+#include <LmbrCentral/Shape/SplineComponentBus.h>
+#include <Source/RigidBodyComponent.h>
+namespace ROS2
```
```suggestion

namespace ROS2
```

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+/*
```
This component should be moved to a separate Gem I believe.

michalpelka replied:
Yes, but I belive it should be work for separate PR.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+                        AZ::Edit::UIHandlers::Default,
+                        &ConveyorBeltComponentKinematic::m_segmentLength,
+                        "Segment Length",
+                        "Length of simulated segments. Short segments causes larger number individual bodies to simulate and challenges "
```
```suggestion
                        "Length of simulated segments. Short segments might affect performance"
```

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+                        &ConveyorBeltComponentKinematic::m_materialAsset,
+                        "Material",
+                        "Material asset of the conveyor belt")
+                    ->Attribute(AZ::Edit::Attributes::DefaultAsset, &GetDefaultPhysicsMaterialAssetId)
```
Is there benefit to using another material asset as default, knowing we are dealing with conveyor belt behavior?

michalpelka replied:
I don understand the question. I think it is beneficial to change the material of the simulated belt.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+
+    void ConveyorBeltComponentKinematic::Activate()
+    {
+        m_initialized = false;
```
Is the field required? 
Could this be replaced with waiting for the entity to fully activate (OnEntityActivated()), then in the handler connecting to TickBus?

michalpelka replied:
Yes.

michalpelka replied:
checked, it works.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+        auto shapeConfiguration =
+            AZStd::make_shared<Physics::BoxShapeConfiguration>(AZ::Vector3(m_segmentLength, m_beltWidth, m_segmentWidth));
+        const auto transform = GetTransformFromSpline(m_splineConsPtr, normalizedLocation);
+        AzPhysics::RigidBodyConfiguration conveyorSegmentRigidBodyConfig;
```
Populating fields of a single segment is a good candidate for a subroutine to help with readibility

michalpelka replied:
True, will add.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+        AzPhysics::SceneInterface* sceneInterface = AZ::Interface<AzPhysics::SceneInterface>::Get();
+        AZ_Assert(sceneInterface, "No scene interface");
+        AzPhysics::SceneHandle defaultSceneHandle = sceneInterface->GetSceneHandle(AzPhysics::DefaultPhysicsSceneName);
+        AZ_Assert(defaultSceneHandle != AzPhysics::InvalidSceneHandle, "Invalid default physics scene handle");
```
Could we instead only subscribe to OnTick when we have all these conditions met?

michalpelka replied:
No. Those are singletons that we should always get.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+        // Strong assumption is that conveyor belt is not transformed after initialization.
+        if (!m_initialized)
+        {
+            m_splineTransform = AZ::Transform::CreateIdentity();
```
This could be moved to subroutine to help with understanding what is happening in onTick

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.cpp:
```diff
@@ -0,0 +1,306 @@
+                &AZ::Render::MaterialComponentRequestBus::Events::SetPropertyValueT<float>,
+                materialId,
+                "uv.offsetU",
+                m_textureOffset);
```
Subroutine MoveBelt could be great here, incorporating spawning, despawning and caching. It could be later replaced with some other behavior (such as moving segments to start) without affecting onTick body.

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.h:
```diff
@@ -0,0 +1,120 @@
+        AZ::Transform m_splineTransform;
+        //! Start and end point of the belt
+        AZ::Vector3 m_startPoint;
+        AZ::Vector3 m_endPoint;
```
Please group members into clearly named structures of configuration.

michalpelka replied:
It is in logical order. please give me more details.

jhanca-robotecai commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.h:
```diff
@@ -0,0 +1,116 @@
+{
+    //! Component that simulates a conveyor belt using kinematic physics.
+    //! The conveyor belt is simulated using a spline and number of kinematic rigid bodies.
+    //! The kinimatic rigid bodies have their kinematic targets set to interpolate along the spline.
```
typo: kinematic

michalpelka commented on Gems/ROS2/Code/CMakeLists.txt:
```diff
@@ -41,8 +41,10 @@ ly_add_target(
             Gem::Atom_RPI.Public
             Gem::Atom_Feature_Common.Static
             Gem::Atom_Component_DebugCamera.Static
+            Gem::AtomLyIntegration_CommonFeatures.Editor.Static
```
It is editor-lib, why it is here?

jhanca-robotecai replied:
This is fixed in my PR

jhanca-robotecai commented:
I created a PR for this one: 
https://github.com/RobotecAI/o3de-extras/pull/24

adamdbrw commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.h:
```diff
@@ -0,0 +1,123 @@
+    //! The conveyor belt is simulated using a spline and number of kinematic rigid bodies.
+    //! The kinematic rigid bodies have their kinematic targets set to interpolate along the spline.
+    //! The component is updating kinematic targets every physic sub-step and creates and despawns rigid bodies as needed.
+    class ConveyorBeltComponentKinematic
```
I suggest changing the name so that the Component is the last part, since this is the convention.
I.e. ConveyorBeltComponent.
I think "Kinematic" is unnecessary.
Also, please add an issue to move FactorySimulation directories to separate Gem.

michalpelka commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentKinematic.h:
```diff
@@ -0,0 +1,123 @@
+        AZ::Vector3 m_startPoint; //!< Start point of the belt
+        AZ::Vector3 m_endPoint; //!< End point of the belt
+        AzPhysics::SceneHandle m_sceneHandle; //!< Scene handle of the scene the belt is in
+        AZStd::string m_graphicalMaterialSlot{ "Belt" }; //!< Name of the material slot to change UVs of
```
I broke the "reflectable part" to separate classes.

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.cpp:
```diff
@@ -0,0 +1,325 @@
+
+        AZ::ConstSplinePtr splinePtr{ nullptr };
+        LmbrCentral::SplineComponentRequestBus::EventResult(splinePtr, m_entity->GetId(), &LmbrCentral::SplineComponentRequests::GetSpline);
+        AZ_Assert(splinePtr, "Spline pointer is null");
```
```suggestion
        AZ_Assert(splinePtr, "Unable to get spline for entity id (%s)", m_entity->GetId().ToString().c_str());
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.cpp:
```diff
@@ -0,0 +1,325 @@
+    AZStd::pair<float, AzPhysics::SimulatedBodyHandle> ConveyorBeltComponent::CreateSegment(
+        AZ::ConstSplinePtr splinePtr, float normalizedLocation)
+    {
+        AzPhysics::SystemInterface* physicsSystem = AZ::Interface<AzPhysics::SystemInterface>::Get();
```
```suggestion
        AzPhysics::SystemInterface* physicsSystem = AZ::Interface<AzPhysics::SystemInterface>::Get();
        AZ_Assert( physicsSystem != nullptr, "Unable to get Physics System");
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.cpp:
```diff
@@ -0,0 +1,325 @@
+            &AZ::Render::MaterialComponentRequestBus::Events::FindMaterialAssignmentId,
+            -1,
+            m_configuration.m_graphicalMaterialSlot);
+
```
Should we validate if a materialId was found here? (like Assert?)

michalpelka replied:
I did not add Assert but decided to add a check before that validate configuration.

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.cpp:
```diff
@@ -0,0 +1,325 @@
+            return;
+        }
+
+        AzPhysics::SceneInterface* sceneInterface = AZ::Interface<AzPhysics::SceneInterface>::Get();
```
```suggestion
        AzPhysics::SceneInterface* sceneInterface = AZ::Interface<AzPhysics::SceneInterface>::Get();
        AZ_Assert( sceneInterface != nullptr, "Unable to get Scene Interface");
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.h:
```diff
@@ -0,0 +1,118 @@
+        //! @return the location of the segment in world space
+        AZ::Vector3 GetLocationOfSegment(const AzPhysics::SimulatedBodyHandle handle);
+
+        //! Obtains the transform of of the pose on the spline at the given distance
```
```suggestion
        //! Obtains the transform of the pose on the spline at the given distance
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponent.h:
```diff
@@ -0,0 +1,118 @@
+        //! @return the transform of the pose on the spline at the given distance
+        AZ::Transform GetTransformFromSpline(AZ::ConstSplinePtr splinePtr, float distanceNormalized);
+
+        //! Spawn rigid body at the given location
```
```suggestion
        //! Spawn a rigid body at the given location
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentConfiguration.cpp:
```diff
@@ -0,0 +1,71 @@
+                        AZ::Edit::UIHandlers::Default,
+                        &ConveyorBeltComponentConfiguration::m_segmentSize,
+                        "Segment Size",
+                        "Size of simulated segments. Short segments might affect performance"
+                        "physics engine.")
```
```suggestion
                        "Size of simulated segments. Short segments might affect performance"
                        "of the physics engine.")
```

spham-amzn commented on Gems/ROS2/Code/Source/FactorySimulation/ConveyorBeltComponentConfiguration.cpp:
```diff
@@ -0,0 +1,71 @@
+                    ->DataElement(
+                        AZ::Edit::UIHandlers::Default,
+                        &ConveyorBeltComponentConfiguration::m_graphicalMaterialSlot,
+                        "Graphical Material slot ",
```
```suggestion
                        "Graphical Material slot",
```

michalpelka commented:
Rebased on top of development and squashed commits. 
Here is backup of reviewed branch:
https://github.com/RobotecAI/o3de-extras/tree/mp/conveyor_belt_backup
